src/libraries/Microsoft.Extensions.Diagnostics/src/Metrics/ListenerSubscription.cs

```diff
@@ -0,0 +1,324 @@
+        {
+            var alreadyEnabled = status.Enabled;
+            var enable = false;
+            var rule = GetMostSpecificRule(instrument);
+            if (rule != null)
+            {
+                enable = rule.Enable;
+            }
+
+            if (!enable && alreadyEnabled)
+            {
+                status.Enabled = false;
+                _meterListener.DisableMeasurementEvents(instrument);
+            }
+            else if (enable && !alreadyEnabled)
+            {
+                // The first time we enable an instrument, we need to call InstrumentPublished.
+                if (!status.Published)
+                {
+                    // However, a listener might decline to enable the instrument, remember that.
+                    status.Published = true;
+                    status.ListenerEnabled = _metricsListener.InstrumentPublished(instrument, out var state);
+                    status.State = state;
+                }
+
+                if (status.ListenerEnabled)
+                {
+                    _meterListener.EnableMeasurementEvents(instrument, status.State);
+                    status.Enabled = true;
+                }
+            }
+        }
+
+        private InstrumentRule? GetMostSpecificRule(Instrument instrument)
+        {
+            InstrumentRule? best = null;
+            foreach (var rule in _rules)
+            {
+                if (RuleMatches(rule, instrument, _metricsListener.Name)
+                    && IsMoreSpecific(rule, best, isLocalScope: instrument.Meter.Scope != null))
+                {
+                    best = rule;
+                }
+            }
+
+            return best;
+        }
+
+        // internal for testing
+        internal static bool RuleMatches(InstrumentRule rule, Instrument instrument, string listenerName)
+        {
+            // Exact match or empty
+            if (!string.IsNullOrEmpty(rule.ListenerName)
+                && !string.Equals(rule.ListenerName, listenerName, StringComparison.OrdinalIgnoreCase))
+            {
+                return false;
+            }
+
+            // Exact match or empty
+            if (!string.IsNullOrEmpty(rule.InstrumentName)
+                && !string.Equals(rule.InstrumentName, instrument.Name, StringComparison.OrdinalIgnoreCase))
+            {
+                return false;
+            }
+
+            if (!(rule.Scopes.HasFlag(MeterScope.Global) && instrument.Meter.Scope == null)
+                && !(rule.Scopes.HasFlag(MeterScope.Local) && instrument.Meter.Scope != null)) // TODO: What should we be comparing Scope to, the DefaultMeterFactory / IMeterFactory?
+            {
+                return false;
+            }
+
+            // Meter
+
+            var ruleMeterName = rule.MeterName.AsSpan();
+            // Rule "System.Net.*" matches meter "System.Net" and "System.Net.Http"
+            if (ruleMeterName.EndsWith(".*".AsSpan(), StringComparison.Ordinal))
```

Comment:
I didn't implement those on purpose, they seem less user friendly and I'm not aware of any scenarios for them. Are you? Consistency isn't enough of a reason without real scenarios.

We support "System.Net" matching that category and all sub categories. Why would you need "System.Net.*" to match only sub categories? The "System.Net.*" syntax seems more logically equivalent to the behavior of "System.Net".

null, "", and "*" all seem logically equivalent. If we support trailing ".*" then I think we should support a root "*" too.

What's the scenario for suffix matching? We wouldn't have many duplicate suffixes that could be matched that way. That's also significantly more complex to implement. Or would someone use it as shorthand for identifying a single meter?

I wasn't aware of the "System*" behavior, I'll revisit that one.

Reply:
> they seem less user friendly and I'm not aware of any scenarios for them

Given that we have years of experience with ILogger to work from, I think we should look at whether users actually complained about the behavior (I'm not aware of any). If we can't find any complaints then I do think consistency should take priority.

For cases we are not aware of the scenario, if the functionality takes 10 lines of code to implement I'd much rather err on the side that we were consistent. I prefer the risk that we added 10 lines for behavior nobody cared about rather than the risk people did have a use we were unaware of and now we broke them and made it inconsistent.